doc/sphinx/language/core/records.rst

```diff
@@ -3,208 +3,300 @@
+      field_spec ::= {* @binder } @of_type
       | {* @binder } := @term
-      term_record ::= %{%| {*; @field_def } {? ; } %|%}
-      field_def ::= @qualid {* @binder } := @term
-
-   Each :n:`@record_definition` defines a record named by :n:`@ident_decl`.
-   The constructor name is given by :n:`@ident`.
-   If the constructor name is not specified, then the default name :n:`Build_@ident` is used,
-   where :n:`@ident` is the record name. If given, the :n:`as @ident`
-   part specifies the name to use for inhabitants of the record in the
-   type of projections (see below).
+      | {* @binder } @of_type := @term
 
-   If :token:`sort` is omitted, the default sort is Type.
-   Notice that the type of an identifier can depend on a previously-given identifier. Thus the
-   order of the fields is important. :n:`@binder` parameters may be applied to the record as a whole
-   or to individual fields.
+   Defines a non-recursive record type, creating projections for each field
+   that has a name other than `_`. The field body and type can depend on previous
+   fields, so the order of fields in the definition may matter.
 
-   .. todo
-      "Record foo2:Prop := { a }." gives the error "Cannot infer this placeholder of type "Type",
-      while "Record foo2:Prop := { a:Type }." gives the output "foo2 is defined.
-      a cannot be defined because it is informative and foo2 is not."
-      Your thoughts?
+   Use the :cmd:`Inductive` and :cmd:`CoInductive` commands to define recursive
+   (inductive or coinductive) records.  These commands also permit defining
+   mutually inductive or coinductive records, provided all of
+   the types in the block are records.  These commands automatically generate
+   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
+   to enable automatic generation of elimination schemes for :cmd:`Record`.
+   See :ref:`proofschemes-induction-principles`.
 
    :n:`{? > }`
-     If provided, the constructor name is automatically declared as
+     If specified, the constructor is declared as
      a coercion from the class of the last field type to the record name
      (this may fail if the uniform inheritance condition is not
-     satisfied).  See :ref:`coercions`.
+     satisfied). See :ref:`coercions`.
+
+   :n:`@ident_decl`
+     The :n:`@ident` within is the name of the record.
+
+   :n:`{* @binder }`
+     :n:`@binder`\s may be used to declare the *parameters* of the record.
+
+   :n:`: @sort`
+     The sort the record belongs to.  The default is :n:`Type`.
+
+   :n:`:= {? @ident }`
+     :n:`@ident` is the name of the record constructor.  If omitted,
+     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.
+
+   :n:`as {? @ident}`
+     Specifies the name used to refer to the argument corresponding to the
+     record in the type of projections.
+
+     .. todo PR: clarify above description, maybe give an example
```

Comment:
Can you help me improve the description, maybe give an example?  I don't understand this.  Thanks.

Reply:
OK, now how do I access or see `HIMOM`?  `About HIMOM` doesn't work.  Nor does `Definition xx := {| bar := 3 |}.  Eval compute in xx.(HIMOM).`

Reply:
`HIMOM` is not a constant. It is a name hint, like other names managed by the `Arguments` command. Typical ways to observe it is as follows:
```
Arguments bar {HIMOM}.
Check  (bar (HIMOM:={|bar:=0|})). 
```

Reply:
Ah, OK.  I added an example, but I don't see how this argument is useful, which would be nice to mention.

Reply:
@herbelin Perhaps you can give me a short explanation of when/why the "as" is useful?  My impression is that it's for convenience.

@Zimmi48 Would you take a look at what we have and assign?  Thanks.

Reply:
> @herbelin Perhaps you can give me a short explanation of when/why the "as" is useful? My impression is that it's for convenience.

Yes, it is to give user control on the name. Virtually, this could be applied also to the name used in the production of induction schemes, or when naming hypotheses of this type in a goal, etc.

Reply:
Ah, the question I meant to ask is why is it beneficial to have HIMOM in `Check  (bar (HIMOM:={|bar:=0|})).` instead of `Check  (bar {|bar:=0|})`?  I see they report different types but I expected they would be equivalent.  Is it just a shortcut so the reduction in Compute isn't needed?  I'm probably misunderstanding something elementary.

The benefit is just "(* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)"?

```
Record foo := { bar : nat ; } as HIMOM.

About bar.
Compute (bar {| bar := 0 |}).   (* = 0 : nat *)
Check (bar {| bar := 0 |}).     (* bar {| bar := 0 |} : nat *)
Arguments bar {HIMOM}.
Check  (bar (HIMOM:={|bar:=0|})).    (* bar : nat *)
Compute (bar {| bar := 0 |}).   (* error (below) *)
```

![image](https://user-images.githubusercontent.com/1253341/153935774-ac8798c2-364c-489e-89e5-d63131883fb6.png)

Reply:
@jfehrle Well that will error because you've made the argument implicit. This shows why it's important to have the name for the variable, because after that argument is implicit there are only two ways of passing that argument. Using the `bar (HIMOM:=_)` syntax or making all implicit arguments explicit with the `@` syntax: `@bar _`.